Enable searching time events in time-graph view

hriday-panchasara opened the pull request:
- updateChart() accepts filter-expressions
- If filter-expressions exist perform SAMPLED fetch followed by DEEP fetch
- DEEP search does a row-by-row fetch

williamsyang-work commented on timeline-chart/src/layer/time-graph-chart.ts:
```diff
@@ -364,8 +366,9 @@ export class TimeGraphChart extends TimeGraphChartLayer {
     }
 
-    updateChart() {
+    async updateChart(filterExpressionsMap?: {[key: number]: string[]}) {
```
Does this need to be async?

bhufmann reviewed (changes requested):
Thanks for this contribution. I have a few comments that need to be addressed.

hriday-panchasara commented on timeline-chart/src/layer/time-graph-chart.ts:
```diff
@@ -497,51 +500,79 @@ export class TimeGraphChart extends TimeGraphChartLayer {
+            if (fullSearch) {
+                for (let i=0;i<rowIds.length;i++) {
+                    const request = { worldRange, resolution, rowIds: [rowIds[i]], additionalParams };
+                    this.fetchRows(request, i === rowIds.length -1, fine, fullSearch);
```
Added await to this.fetchRows call within the for loop

hriday-panchasara commented on timeline-chart/src/layer/time-graph-chart.ts:
```diff
@@ -497,51 +500,79 @@ export class TimeGraphChart extends TimeGraphChartLayer {
-                if (!fine && this._coarseResolutionFactor !== FINE_RESOLUTION_FACTOR) {
-                    this._debouncedMaybeFetchNewDataFine();
+            if (fullSearch) {
+                for (let i=0;i<rowIds.length;i++) {
```
@bhufmann I'm not sure if I've added the correct conditional to break out of the for loop. My assumption is as follows:

updateChart() calls w/ filter:
1st) Sampled search Coarse render (all rows)
2nd) Sampled search Fine render (all rows)
3rd) Deep search fine render (row-by-row)

updateChart() calls w/o filter:
1st) Coarse render (all rows)
2nd) Fine render (all rows)

Based on the above, I'm assuming that row-by-row fetch should be cancelled if another call is made to updateChart with Coarse render (all rows).

bhufmann reviewed (changes requested):
Thanks for the updates. I have some comments below though.

bhufmann commented on timeline-chart/src/layer/time-graph-chart.ts:
```diff
@@ -497,51 +500,83 @@ export class TimeGraphChart extends TimeGraphChartLayer {
-                        this.updateZoomingSelection();
+            if (fullSearch) {
+                for (let i = 0; i < rowIds.length; i++) {
+                    if (this.ongoingRequest && !this.ongoingRequest.fullSearch) {
```
Hmm, I'm not able to reproduce this issues. So, ignore this comment for now. I'll keep an eye on it when testing your next revision of this PR.

bhufmann commented on timeline-chart/src/layer/time-graph-chart.ts:
```diff
@@ -497,51 +500,83 @@ export class TimeGraphChart extends TimeGraphChartLayer {
-                return;
+            const additionalParams: { [key: string]: any } = {};
+            if (this._filterExpressionsMap) {
+                additionalParams['filter_query_parameters'] = {'filter_expressions_map': this._filterExpressionsMap, 'strategy': fullSearch ? 'DEEP' : 'SAMPLED'};
```
@hriday-panchasara I tried to come-up with a proposal that makes this less coupled to the `theia-trace-extension` implementation. However, it's not straight forward due the need to pass the strategy back. If you meanwhile have found a way please let me know.

So, I opened a ticket so that we can address the tight coupling in a later effort #256. My suggestion going forward is to keep this implementation as proposed in the PR (you can add a comment to the new APIs about #256). Afterwards, we can work on the refactoring of the implementation when fixing #256. We can then take our time to come-up with a better, decoupled API and implementation. WDYT?

hriday-panchasara replied:
Thanks for the suggestion. I'm also struggling to take out the `filter_query_parameters`, mostly because the current implementation is tied with the changes brought by #180 & #205. Therefore, decoupling from this point would affect current functionality. My suggestion also would be to use #256 to roll back and find a solution that satisfies all the functional requirements and performance optimizations that exist within this library.

bhufmann reviewed (changes requested):
@hriday-panchasara please rebase the change to the latest in main branch so that I can continue reviewing and testing it.

hriday-panchasara commented:
> I tested it with the corresponding patch ([eclipse-cdt-cloud/theia-trace-extension#886](https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/886)) and I found 2 issues. For the first one, I can't share screenshot here but I'll sent it to you separately.
> 
> * For a large trace with a lot of states per row and fully zoomed out, I entered a search string that has many matches. The view will update row by rows slowly. While it's ongoing I changed the search string to a string without many matches, then the previous search is supposed to be cancelled and the new one should be applied. What I see here, that rows are still updated for the first search string.
> * I have a trace with lost events. I entered a search string which updates the view with the matches. Then I use the toolbar menu to hide lost events, which causes the view to update, but the search matches are not applied (see below).
> 
> ![tg-filter-issue-markers](https://user-images.githubusercontent.com/413424/242296247-a73fd4f5-7ab5-4116-8cb2-e27e6e260dde.gif)

Thanks for the feedback Bernd. I've updated both patches to introduce the following changes:
- Cancel row-by-row full search if a new request is made to /states
- Skip full search for rows that already have the correct model
- Filter applied to /states call made when markers/collapsedNodes is updated

bhufmann reviewed (changes requested):
Thinks work better but still there seems to be another corner case:
- pressing the zoom button or resizing window doesn't update all the rows properly, only the deep search at the end fixes the display

hriday-panchasara commented:
> Thinks work better but still there seems to be another corner case:
> 
> * pressing the zoom button or resizing window doesn't update all the rows properly, only the deep search at the end fixes the display

Added a new patch to resolve issues with zoom. Explanation of changes:
- Cancel row-by-row full search whenever a new call is being made to maybeFetchNewData through updates to viewRange/rowIds/zoomRange/resolution/ markers/filters. If row-by-row fetch is successful (not interrupted by a new onGoingRequest), update the rowComponent models with the filterExpressionsMap.
- It is an inefficient approach since we refetch all states for even small changes, but its still a better user experience than calling deep search on all rows at once. We can work on optimizing row-by-row full search in subsequent patches.

bhufmann reviewed (approved):
The feature works well. It is able to apply the results correctly. If there are many zooms operations performed short after each other, there are some update glitches, due the row by row updates for the filters. The end result, however, is correct and the view is drawn correctly. I think at this point we can go forward with this change as-is.  I'll add a follow-up task to investigate update glitches. I asked @PatrickTasse to help with this investigation.
